---
title: Wyjątki i obsługa wyjątków — przewodnik programowania c#
ms.date: 07/20/2015
helpviewer_keywords:
- exception handling [C#]
- exceptions [C#]
- C# language, exceptions
ms.assetid: 0001887f-4fa2-47e2-8034-2819477e2344
ms.openlocfilehash: b883012cf8f72247ff4e0b47a46eee1854e2d534
ms.sourcegitcommit: 7588136e355e10cbc2582f389c90c127363c02a5
ms.translationtype: MT
ms.contentlocale: pl-PL
ms.lasthandoff: 03/15/2020
ms.locfileid: "76735649"
---
# <a name="exceptions-and-exception-handling-c-programming-guide"></a><span data-ttu-id="28c33-102">Wyjątki i obsługa wyjątków (Przewodnik programowania w języku C#)</span><span class="sxs-lookup"><span data-stu-id="28c33-102">Exceptions and Exception Handling (C# Programming Guide)</span></span>

<span data-ttu-id="28c33-103">Funkcje obsługi wyjątków języka C# ułatwiają radzenie sobie z nieoczekiwanymi lub wyjątkowymi sytuacjami, które występują, gdy program jest uruchomiony.</span><span class="sxs-lookup"><span data-stu-id="28c33-103">The C# language's exception handling features help you deal with any unexpected or exceptional situations that occur when a program is running.</span></span> <span data-ttu-id="28c33-104">Obsługa wyjątków `try` `catch`używa `finally` , i słów kluczowych, aby wypróbować akcje, które mogą się nie powieść, do obsługi błędów, gdy zdecydujesz, że jest to uzasadnione, a następnie oczyścić zasoby.</span><span class="sxs-lookup"><span data-stu-id="28c33-104">Exception handling uses the `try`, `catch`, and `finally` keywords to try actions that may not succeed, to handle failures when you decide that it is reasonable to do so, and to clean up resources afterward.</span></span> <span data-ttu-id="28c33-105">Wyjątki mogą być generowane przez program runtime języka wspólnego (CLR), przez .NET Framework lub biblioteki innych firm lub przez kod aplikacji.</span><span class="sxs-lookup"><span data-stu-id="28c33-105">Exceptions can be generated by the common language runtime (CLR), by the .NET Framework or any third-party libraries, or by application code.</span></span> <span data-ttu-id="28c33-106">Wyjątki są tworzone przy `throw` użyciu słowa kluczowego.</span><span class="sxs-lookup"><span data-stu-id="28c33-106">Exceptions are created by using the `throw` keyword.</span></span>

<span data-ttu-id="28c33-107">W wielu przypadkach wyjątek może być generowany nie przez metodę, która została wywołana bezpośrednio przez kod, ale przez inną metodę w dół w stosie wywołań.</span><span class="sxs-lookup"><span data-stu-id="28c33-107">In many cases, an exception may be thrown not by a method that your code has called directly, but by another method further down in the call stack.</span></span> <span data-ttu-id="28c33-108">W takim przypadku CLR będzie odwijać stosu, szukając metody z blokiem `catch` `catch` dla określonego typu wyjątku i wykona pierwszy taki blok, który w przypadku znalezienia.</span><span class="sxs-lookup"><span data-stu-id="28c33-108">When this happens, the CLR will unwind the stack, looking for a method with a `catch` block for the specific exception type, and it will execute the first such `catch` block that if finds.</span></span> <span data-ttu-id="28c33-109">Jeśli nie znajdzie `catch` odpowiedniego bloku w dowolnym miejscu w stosie wywołań, zakończy proces i wyświetli komunikat do użytkownika.</span><span class="sxs-lookup"><span data-stu-id="28c33-109">If it finds no appropriate `catch` block anywhere in the call stack, it will terminate the process and display a message to the user.</span></span>

<span data-ttu-id="28c33-110">W tym przykładzie metoda testuje podział przez zero i przechwytuje błąd.</span><span class="sxs-lookup"><span data-stu-id="28c33-110">In this example, a method tests for division by zero and catches the error.</span></span> <span data-ttu-id="28c33-111">Bez obsługi wyjątków ten program zakończy się z **DivideByZeroException był nieobsługiwany** błąd.</span><span class="sxs-lookup"><span data-stu-id="28c33-111">Without the exception handling, this program would terminate with a **DivideByZeroException was unhandled** error.</span></span>

[!code-csharp[csProgGuideExceptions#18](~/samples/snippets/csharp/VS_Snippets_VBCSharp/csProgGuideExceptions/CS/Exceptions.cs#18)]

## <a name="exceptions-overview"></a><span data-ttu-id="28c33-112">Przegląd wyjątków</span><span class="sxs-lookup"><span data-stu-id="28c33-112">Exceptions Overview</span></span>

<span data-ttu-id="28c33-113">Wyjątki mają następujące właściwości:</span><span class="sxs-lookup"><span data-stu-id="28c33-113">Exceptions have the following properties:</span></span>

- <span data-ttu-id="28c33-114">Wyjątki to typy, które ostatecznie `System.Exception`pochodzą z .</span><span class="sxs-lookup"><span data-stu-id="28c33-114">Exceptions are types that all ultimately derive from `System.Exception`.</span></span>
- <span data-ttu-id="28c33-115">Użyj `try` bloku wokół instrukcji, które mogą zgłaszać wyjątki.</span><span class="sxs-lookup"><span data-stu-id="28c33-115">Use a `try` block around the statements that might throw exceptions.</span></span>
- <span data-ttu-id="28c33-116">Gdy wystąpi wyjątek `try` w bloku, przepływ kontroli skacze do pierwszego skojarzonego programu obsługi wyjątków, który jest obecny w dowolnym miejscu w stosie wywołań.</span><span class="sxs-lookup"><span data-stu-id="28c33-116">Once an exception occurs in the `try` block, the flow of control jumps to the first associated exception handler that is present anywhere in the call stack.</span></span> <span data-ttu-id="28c33-117">W języku `catch` C#słowo kluczowe służy do definiowania programu obsługi wyjątków.</span><span class="sxs-lookup"><span data-stu-id="28c33-117">In C#, the `catch` keyword is used to define an exception handler.</span></span>
- <span data-ttu-id="28c33-118">Jeśli nie jest program obsługi wyjątków dla danego wyjątku, program przestaje wykonywać z komunikatem o błędzie.</span><span class="sxs-lookup"><span data-stu-id="28c33-118">If no exception handler for a given exception is present, the program stops executing with an error message.</span></span>
- <span data-ttu-id="28c33-119">Nie łapiesz wyjątku, chyba że można go obsłużyć i pozostawić aplikację w znanym stanie.</span><span class="sxs-lookup"><span data-stu-id="28c33-119">Do not catch an exception unless you can handle it and leave the application in a known state.</span></span> <span data-ttu-id="28c33-120">Jeśli złapiesz `System.Exception`, ponownie `throw` wrzuć go za `catch` pomocą słowa kluczowego na końcu bloku.</span><span class="sxs-lookup"><span data-stu-id="28c33-120">If you catch `System.Exception`, re-throw it using the `throw` keyword at the end of the `catch` block.</span></span>
- <span data-ttu-id="28c33-121">Jeśli `catch` blok definiuje zmienną wyjątku, można go użyć, aby uzyskać więcej informacji na temat typu wyjątku, który wystąpił.</span><span class="sxs-lookup"><span data-stu-id="28c33-121">If a `catch` block defines an exception variable, you can use it to obtain more information about the type of exception that occurred.</span></span>
- <span data-ttu-id="28c33-122">Wyjątki mogą być jawnie generowane przez `throw` program przy użyciu słowa kluczowego.</span><span class="sxs-lookup"><span data-stu-id="28c33-122">Exceptions can be explicitly generated by a program by using the `throw` keyword.</span></span>
- <span data-ttu-id="28c33-123">Obiekty wyjątku zawierają szczegółowe informacje o błędzie, takie jak stan stosu wywołań i opis tekstowy błędu.</span><span class="sxs-lookup"><span data-stu-id="28c33-123">Exception objects contain detailed information about the error, such as the state of the call stack and a text description of the error.</span></span>
- <span data-ttu-id="28c33-124">Kod w `finally` bloku jest wykonywany, nawet jeśli wyjątek.</span><span class="sxs-lookup"><span data-stu-id="28c33-124">Code in a `finally` block is executed even if an exception is thrown.</span></span> <span data-ttu-id="28c33-125">Użyj `finally` bloku, aby zwolnić zasoby, na przykład, aby zamknąć `try` wszystkie strumienie lub pliki, które zostały otwarte w bloku.</span><span class="sxs-lookup"><span data-stu-id="28c33-125">Use a `finally` block to release resources, for example to close any streams or files that were opened in the `try` block.</span></span>
- <span data-ttu-id="28c33-126">Wyjątki zarządzane w platformie .NET Framework są implementowane na podstawie mechanizmu obsługi wyjątków strukturalnych Win32.</span><span class="sxs-lookup"><span data-stu-id="28c33-126">Managed exceptions in the .NET Framework are implemented on top of the Win32 structured exception handling mechanism.</span></span> <span data-ttu-id="28c33-127">Aby uzyskać więcej informacji, zobacz [Obsługa wyjątków strukturalnych (C/C++)](/cpp/cpp/structured-exception-handling-c-cpp) i [Kurs awarii na głębokości win32 obsługa wyjątków strukturalnych](http://bytepointer.com/resources/pietrek_crash_course_depths_of_win32_seh.htm).</span><span class="sxs-lookup"><span data-stu-id="28c33-127">For more information, see [Structured Exception Handling (C/C++)](/cpp/cpp/structured-exception-handling-c-cpp) and [A Crash Course on the Depths of Win32 Structured Exception Handling](http://bytepointer.com/resources/pietrek_crash_course_depths_of_win32_seh.htm).</span></span>

## <a name="related-sections"></a><span data-ttu-id="28c33-128">Sekcje pokrewne</span><span class="sxs-lookup"><span data-stu-id="28c33-128">Related Sections</span></span>

<span data-ttu-id="28c33-129">Zobacz następujące artykuły, aby uzyskać więcej informacji na temat wyjątków i obsługi wyjątków:</span><span class="sxs-lookup"><span data-stu-id="28c33-129">See the following articles for more information about exceptions and exception handling:</span></span>

- [<span data-ttu-id="28c33-130">Używanie wyjątków</span><span class="sxs-lookup"><span data-stu-id="28c33-130">Using Exceptions</span></span>](using-exceptions.md)
- [<span data-ttu-id="28c33-131">Obsługa wyjątków</span><span class="sxs-lookup"><span data-stu-id="28c33-131">Exception Handling</span></span>](exception-handling.md)
- [<span data-ttu-id="28c33-132">Tworzenie i zgłaszanie wyjątków</span><span class="sxs-lookup"><span data-stu-id="28c33-132">Creating and Throwing Exceptions</span></span>](creating-and-throwing-exceptions.md)
- [<span data-ttu-id="28c33-133">Wyjątki generowane przez kompilator</span><span class="sxs-lookup"><span data-stu-id="28c33-133">Compiler-Generated Exceptions</span></span>](compiler-generated-exceptions.md)
- [<span data-ttu-id="28c33-134">Jak obsługiwać wyjątek przy użyciu try/catch (Przewodnik programowania C#)</span><span class="sxs-lookup"><span data-stu-id="28c33-134">How to handle an exception using try/catch (C# Programming Guide)</span></span>](how-to-handle-an-exception-using-try-catch.md)
- [<span data-ttu-id="28c33-135">Wykonywanie czyszczenia kodu za pomocą instrukcji finally</span><span class="sxs-lookup"><span data-stu-id="28c33-135">How to execute cleanup code using finally</span></span>](how-to-execute-cleanup-code-using-finally.md)
- [<span data-ttu-id="28c33-136">Przechwytywanie wyjątku typu non-CLS</span><span class="sxs-lookup"><span data-stu-id="28c33-136">How to catch a non-CLS exception</span></span>](how-to-catch-a-non-cls-exception.md)

## <a name="c-language-specification"></a><span data-ttu-id="28c33-137">Specyfikacja języka C#</span><span class="sxs-lookup"><span data-stu-id="28c33-137">C# Language Specification</span></span>

<span data-ttu-id="28c33-138">Aby uzyskać więcej informacji, zobacz [wyjątki](~/_csharplang/spec/exceptions.md) w [specyfikacji języka Języka C#](/dotnet/csharp/language-reference/language-specification/introduction).</span><span class="sxs-lookup"><span data-stu-id="28c33-138">For more information, see [Exceptions](~/_csharplang/spec/exceptions.md) in the [C# Language Specification](/dotnet/csharp/language-reference/language-specification/introduction).</span></span> <span data-ttu-id="28c33-139">Specyfikacja języka jest ostatecznym źródłem informacji o składni i użyciu języka C#.</span><span class="sxs-lookup"><span data-stu-id="28c33-139">The language specification is the definitive source for C# syntax and usage.</span></span>

## <a name="see-also"></a><span data-ttu-id="28c33-140">Zobacz też</span><span class="sxs-lookup"><span data-stu-id="28c33-140">See also</span></span>

- <xref:System.SystemException>
- [<span data-ttu-id="28c33-141">Przewodnik programowania języka C#</span><span class="sxs-lookup"><span data-stu-id="28c33-141">C# Programming Guide</span></span>](../index.md)
- [<span data-ttu-id="28c33-142">Słowa kluczowe języka C#</span><span class="sxs-lookup"><span data-stu-id="28c33-142">C# Keywords</span></span>](../../language-reference/keywords/index.md)
- [<span data-ttu-id="28c33-143">throw</span><span class="sxs-lookup"><span data-stu-id="28c33-143">throw</span></span>](../../language-reference/keywords/throw.md)
- [<span data-ttu-id="28c33-144">try-catch</span><span class="sxs-lookup"><span data-stu-id="28c33-144">try-catch</span></span>](../../language-reference/keywords/try-catch.md)
- [<span data-ttu-id="28c33-145">try-finally</span><span class="sxs-lookup"><span data-stu-id="28c33-145">try-finally</span></span>](../../language-reference/keywords/try-finally.md)
- [<span data-ttu-id="28c33-146">try-catch-finally</span><span class="sxs-lookup"><span data-stu-id="28c33-146">try-catch-finally</span></span>](../../language-reference/keywords/try-catch-finally.md)
- [<span data-ttu-id="28c33-147">Wyjątki</span><span class="sxs-lookup"><span data-stu-id="28c33-147">Exceptions</span></span>](../../../standard/exceptions/index.md)
